apps/prairielearn/src/ee/auth/azure/index.ts

```diff
@@ -13,7 +13,14 @@ export function getAzureStrategy() {
       cookieEncryptionKeys: config.azureCookieEncryptionKeys,
       loggingLevel: config.azureLoggingLevel,
       scope: ['openid', 'profile', 'email'],
-      responseType: 'code id_token',
+      // This should only be `id_token`, which will force the implicit flow to
+      // be used. When we specific `code id_token` here, it will use the hybrid
+      // flow, which for whatever reason does not return user emails.
+      //
+      // The implicit flow isn't recommended for `response_type=token`, but it's
+      // safe for us to use here because we're using `response_type=id_token`. We
+      // don't ever need access/refresh tokens for our use case.
+      responseType: 'id_token',
```

Comment:
This change was crucial to get things working. When using the hybrid flow, all the actual profile information comes from requesting it from the token endpoint (https://login.microsoftonline.com/common/oauth2/token):

https://github.com/AzureAD/passport-azure-ad/blob/0fd15d1e1dac2ed82c52c93e95e802f5a586b7d5/lib/oidcstrategy.js#L1143-L1173

For whatever reason, that token does _not_ include the email, even through the `id_token` in the initial response did.

Interestingly, the "v2" version of Microsoft's OAuth endpoints (https://login.microsoftonline.com/common/v2.0/.well-known/openid-configuration) _does_ return email, and the hybrid flow actually completely skips trying to fetch anything from the "userinfo" endpoint:

https://github.com/AzureAD/passport-azure-ad/blob/0fd15d1e1dac2ed82c52c93e95e802f5a586b7d5/lib/oidcstrategy.js#L1318

However, we can't use this v2 version because it doesn't contain `upn` 😢